### PR TITLE
[main] moe(perf): Pre-GDR kernel fusion

### PR DESCRIPTION
**Pull request by yuzhongw-nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

PR for dev: https://github.com/NVIDIA/Megatron-LM/pull/5361

This PR adds a streamed pre-gated-delta-rule fusion path for GatedDeltaNet.

The new path fuses the pre-GDR preprocessing work, including causal convolution, layout transforms, QK normalization, value/gate preparation, and beta/decay computation. It is exposed through a boolean TransformerConfig option, gdn_pre_gated_delta_rule_fusion, and defaults to disabled.

Based on Qwen3.5-397B-A17B shaped GDN and MBS=4, estimated ~3x speedup for the fused scope and ~1.05x for proxy model E2E training.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/fusions/fused_pre_gated_delta_rule.py [hunk 1]
```diff
@@ -0,0 +1,1778 @@
...
+    alpha = tl.load(alpha_ptr, mask=mask, other=0.0).to(tl.float32)
+    beta = tl.load(beta_ptr, mask=mask, other=0.0).to(tl.float32)
+
+    A_log = tl.load(A_log_ptr + h_offs, mask=h_mask, other=0.0).to(tl.float32)
+    dt_bias = tl.load(dt_bias_ptr + h_offs, mask=h_mask, other=0.0).to(tl.float32)
+
+    pre = alpha + dt_bias[None, :]
+    # softplus(x) = log(1 + exp(x)); torch's softplus thresholds at x>20 but we
+    # rely on fp32 evaluation here, which stays well within range for typical
+    # GDN inputs (the unfused path computes the same expression).
```
Comment (Mellonta) [inline megatron/core/fusions/fused_pre_gated_delta_rule.py]: Is this assumption safe? In FP32, exp(x) overflows around x = 89.
Shall we do
```
@triton.jit
def softplus(x):
    return tl.where(x < 20.0, tl.math.log(1 + tl.math.exp(x)), x)
```
to match torch's numerics?

File: megatron/core/fusions/fused_pre_gated_delta_rule.py [hunk 2]
```diff
@@ -0,0 +1,1778 @@
...
+        # No l2norm follows. The final store→bf16 already does the rounding;
+        # an intermediate bf16 round-trip would be redundant.
+        out = silu_out
+
+    out_typed = out.to(out_ptr.dtype.element_ty)
+
+    # Write the same data to ``REPEAT`` adjacent value heads. ``REPEAT == 1``
+    # is the no-repeat case (V branch is handled by a separate kernel that
+    # always has REPEAT == 1, but using the same code here is convenient).
+    for r in tl.static_range(REPEAT):
```
Comment (Mellonta) [inline megatron/core/fusions/fused_pre_gated_delta_rule.py:221]: Could we avoid materializing repeated Q/K heads here? FLA keeps Q/K at H heads and maps each value head with qk_head = value_head /  (HV / H). Is equal head count required downstream?

File: tests/unit_tests/ssm/test_gated_delta_net.py [hunk 3]
```diff
@@ -412,6 +463,493 @@ def test_gpu_forward_thd_padding_correctness(self):
...
 
+@pytest.mark.skipif(not HAVE_FLA, reason="FLA is not installed.")
+@pytest.mark.skipif(not HAVE_FUSED_PRE_GDR, reason="causal-conv1d fused backward is not installed.")
+@pytest.mark.internal
+class TestFusedPreGatedDeltaRule:
+
+    @pytest.fixture(scope='function', autouse=True)
+    def setup_method(self):
+        Utils.initialize_model_parallel(
+            tensor_model_parallel_size=1, pipeline_model_parallel_size=1, context_parallel_size=1
```
Comment (Mellonta) [inline tests/unit_tests/ssm/test_gated_delta_net.py:679]: Does the fused kernel support CP?

File: megatron/core/ssm/gated_delta_net.py [hunk 4]
```diff
@@ -385,6 +392,89 @@ def forward(
...
+                else None
+            )
+            query, key, value, gate, beta, g = self._fused_streamed_pre_gated_delta_rule(
+                qkvzba, cu_seqlens_q=cu_seqlens_q, seq_idx=seq_idx
+            )
+            nvtx_range_pop(suffix="fused_streamed_pre_gated_delta_rule")
+        else:
+            nvtx_range_push(suffix="pre_gated_delta_rule")
+            query, key, value, gate, beta, g = self.pre_gated_delta_rule(
+                qkvzba, batch, seq_len, cp_size, cp_group, cu_seqlens_q
```
Comment (Mellonta) [inline megatron/core/ssm/gated_delta_net.py]: cp_size should be self.cp_size?

File: tests/unit_tests/ssm/test_gated_delta_net.py (hunk 3 above)
Comment (yuzhongw-nvidia) [reply tests/unit_tests/ssm/test_gated_delta_net.py:679]: Not yet. I'm currently working on it (https://github.com/NVIDIA/Megatron-LM/pull/5638). I'll let you know when I finish it.

File: megatron/core/fusions/fused_pre_gated_delta_rule.py (hunk 2 above)
Comment (yuzhongw-nvidia) [reply megatron/core/fusions/fused_pre_gated_delta_rule.py:221]: I remember FLA does not support what you said in the past, but let me check it again. If so, it will help a lot. Thanks for your reminder.

**Review by yaoyu-33 — APPROVED** [review]

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/6dbc8a97c6097fe4ccdaa98fc92452e76091b674

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5532/commits/27e8cc7455004dc5f716de25bdfcb60c1cc8794f

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/10dacdf7a0925179ad8afde3a087926da3b981da

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/2e577b58975e6b0a4ca008ada2705cd4e84e1f84

File: megatron/core/fusions/fused_pre_gated_delta_rule.py [hunk 5]
```diff
@@ -0,0 +1,1711 @@
...
+) -> Optional[Tensor]:
+    """Return the token-level sequence-id buffer for causal-conv backward."""
+
+    if cu_seqlens is None:
+        assert seq_idx is None, "seq_idx requires cu_seqlens for packed THD mode."
+        return None
+
+    if seq_idx is None:
+        seq_lengths = cu_seqlens[1:] - cu_seqlens[:-1]
+        seq_idx = torch.repeat_interleave(
```
Comment (CarlosGomes98) [inline megatron/core/fusions/fused_pre_gated_delta_rule.py:1060]: we should provide output_size= to avoid a sync point
https://docs.pytorch.org/docs/2.13/generated/torch.repeat_interleave.html

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5532/commits/75746c430925818ad848339b8863d7a8e7aade29

File: megatron/core/fusions/fused_pre_gated_delta_rule.py (hunk 5 above)
Comment (yuzhongw-nvidia) [reply megatron/core/fusions/fused_pre_gated_delta_rule.py:1060]: Fixed. Thanks!

**Review by FDecaYed — APPROVED** [review]

**Review by zhongbozhu — APPROVED** [review]

File: codecov.yml [hunk 6]
```diff
@@ -12,3 +12,5 @@ coverage:
         if_not_found: success
 fixes:
   - "/opt/megatron-lm/::"
+ignore:
+  - "megatron/core/fusions/fused_pre_gated_delta_rule.py"
```
Comment (Phlip79) [inline codecov.yml]: Don't add this, the change to pyproject.toml should be enough

File: pyproject.toml [hunk 7]
```diff
@@ -296,6 +296,10 @@ omit = [
     "/opt/megatron-lm/tools/*",
     "/usr/local/lib/python3.12/dist-packages/*",
     "/opt/megatron-lm/_remote_module_non_scriptable",
+    # These fused paths are mostly Triton DSL kernels and PyTorch
+    # autograd-engine callbacks. The CUDA code is exercised by unit tests,
+    # but coverage.py cannot reliably trace those execution contexts.
+    "*/megatron/core/fusions/*",
```
Comment (Phlip79) [inline pyproject.toml]:
```suggestion
    "/opt/megatron-lm/megatron/core/fusions/*",
```

**Review by guihong-nv — APPROVED** [review]

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/fb693b5715db09b08e59b908bd38e3cc0a5f843d

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/079ceecd9a04d124dd5e77554789a03ea57676af

**Review by ko3n1g — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
